### DISCUSSION TITLE
"webServiceUrl" is not getting called by Apple Wallet. Not returning deviceLibraryIdentifier.

### DISCUSSION THREAD

**Discussion by ahmadaIanazi** [alexandercerutti/passkit-generator#134] [Q&A] [answered]
When the end-user taps on "Add to wallet" and adds the Pass to their Apple Wallet, there should be some magic happening behind the scene, where Apple server sends back and registers the device, by calling this endpoint:

`/v1/passes/pass.com.example.MyPassType/ABC123`

But nothing is happening, I have tried a lot of time, HTTP + HTTPS, but I am not sure about the "SSL certificates where I should get a specific one or what" Does anyone have a faced any issues in that regards, or at least have successfully register a device. 

`"webServiceUrl": "http://app.bettaga.com" 
`
Thank you very much.

**Comment by alexandercerutti** [comment, marked as answer]
Hi there @ahmadaIanazi, thank you for using passkit-generator!

Do you mind reporting where you read about the SSL Certificate? I've never implemented a server to update a pass yet, but I'm curious.

**Reply by ahmadaIanazi** [reply]
Hi Alex, I am not quit sure about the SSL, its mentioned in Apple Docs, that in the development mode, you can use HTTP but in production it must be an HTTPS, either ways didnt work. I am trying to debug why the server is not being called. Have you ever got the **deviceLibraryIdentifier** ? This is where it all starts. This will help us update a pass, and sends notification to this specific pass ..etc

Is there another way to update, send push notification to a certain pass ? Please let me know if you know. 

Thank you,

**Reply by alexandercerutti** [reply]
For the HTTP Mode, there's a switch in the Developer Settings on your iPhone.
I honestly don't know about HTTPS.

About the server not being called, I honestly don't know. Have you tried to set up [the log endpoint](https://developer.apple.com/documentation/walletpasses/log_a_message) to check if something meaningful gets logged?

> Have you ever got the deviceLibraryIdentifier ? This is where it all starts. This will help us update a pass, and sends notification to this specific pass ..etc

I think you get it when the server receives the request of this endpoint: https://developer.apple.com/documentation/walletpasses/register_a_pass_for_update_notifications

Let me know

**Reply by ahmadaIanazi** [reply]
Hi Alex, we have finally figured it out. It was a silly but very confusing error. 

1. Serial number provided with the pass.json must be unique every-time even during testing, because it will not be called if it existed before.
2. "webServiceURL" must be written exactly like that. 
3. The Web Service URL must end with a forward slash. 

Good luck.

**Reply by alexandercerutti** [reply]
Great to know, @ahmadaIanazi! Thank you for sharing your discoveries!

**Reply by prashant03Dev** [reply]
@alexandercerutti  How do we install the wallet pass to the device using webservice url,and how do we get the webServiceUrl?

**Reply by alexandercerutti** [reply]
@prashant03Dev what? You should already know your webServiceUrl when you generate the pass. Let's use a different discussion, if you have further questions.

**Reply by Bekshakx** [reply]
@ahmadaIanazi  hi, i have the same problem. can you help me? look at my code and advise?

**Reply by alexandercerutti** [reply]
@Bekshakx open a new issue with all your details. I'll try to help you as I can.

**Reply by videomat** [reply]
> Hi Alex, we have finally figured it out. It was a silly but very confusing error.
> 
> 1. Serial number provided with the pass.json must be unique every-time even during testing, because it will not be called if it existed before.
> 2. "webServiceURL" must be written exactly like that.
> 3. The Web Service URL must end with a forward slash.
> 
> Good luck.

Just an additional comment that even though Apple says HTTP is fine for development, their background POST requests still seem to require HTTPS.
I tried setting `webServiceURL` to my local IP (like http://192.168.0.12/). That broke Wallet pass generation entirely. Switching to HTTPS allowed the pass to be generated, but the device still didn’t hit the /v1/devices/ registration endpoint.

Using Ngrok to tunnel my local server over HTTPS finally solved it.
Hope this saves someone a few hours.

**Reply by alexandercerutti** [reply]
@videomat in order to unlock support to http, you need to have the option set in you development iPhone. In the Developer menu, in Settings, there's a dedicated option called "Allow HTTP services" (or something like that, I'm translating literally from Italian).

<img src="https://github.com/user-attachments/assets/de48bce8-a9af-4947-bd21-6eaef4ce050e" width="300" />

**Reply by videomat** [reply]
Thanks! I totally missed your comment above on this one :)
